crates/subspace-archiving/src/archiver.rs

```diff
@@ -356,23 +363,33 @@ impl Archiver {
         object_mapping: BlockObjectMapping,
         incremental: bool,
     ) -> ArchiveBlockOutcome {
+        let mut archived_segments = Vec::new();
+        let mut global_object_mapping = Vec::new();
+
+        // Since mappings are produced immediately after the block is added, we can assume that
+        // the mappings from any with_initial_state() block have already been produced.
```

Comment:
This comment is in seemingly arbitrary location and I'm not sure it provides useful context since `add_block` is a public method that has really nothing to do with `with_initial_state()` method.

I read all commits and I think this comment is somehow related to the fact that you removed `object_mapping` argument from the `with_initial_state()` constructor, but I am puzzled why you removed it there in the first place :thinking:

Reply:
The comment in `with_initial_state()` explains why it was removed: we already produced mappings for that whole block the first time any part of it was archived. (And are unable to produce mappings inside `with_initial_state()`, because if the remainder of the block is bigger than a segment, we’d have to run the whole archiving machinery to make sure the mappings are in the correct segments.)

So if any comment needs work, let’s change that one?
https://github.com/autonomys/subspace/pull/3100/files#diff-fa710a837f7e5aaaea197e51ee425eaad246333c7a343fbd38b6754f789f3edaR326

Happy to delete this one.

```suggestion
```

Reply:
> The comment in with_initial_state() explains why it was removed: we already produced mappings for that whole block the first time any part of it was archived

Except when node shutdown happened shortly after it was archived. It will probably not make much of a difference for RPC since the client is unlikely to reconnect that fast, but when using these crates as a library it will make a difference.

Reply:
Ok, we can do that. What do you think of commit [3b7df06](https://github.com/autonomys/subspace/pull/3100/commits/3b7df06bde0c765787af6540b5d5f8d1063d0bd3)?

Reply:
I wouldn't produce object mappings on initialization. We can delay this to when the next block is processed by making `produce_object_mappings` process all segment items in a loop similarly to how it did that before this PR. The public API before this PR was already good, I don't think we need to change it and returning `ArchiveBlockOutcome` from the constructor is quite awkward.